src/trace/snapshotter.ts

```diff
@@ -0,0 +1,525 @@
+        if (childSnapshot)
+          result.childFrames.push({ frameSrc, snapshot: childSnapshot });
+      }
+
+      return result;
+    } catch (e) {
+      return null;
+    }
+  };
+
+  const mainFrameSnapshot = await visit(page.mainFrame());
+  if (!mainFrameSnapshot)
+    return null;
+
+  let viewportSize = page.viewportSize();
+  if (!viewportSize) {
+    try {
+      if (!progress.isRunning())
+        return null;
+
+      const context = await page.mainFrame()._utilityContext();
+      viewportSize = await context.evaluateInternal(() => {
+        return {
+          width: Math.max(document.body.offsetWidth, document.documentElement.offsetWidth),
+          height: Math.max(document.body.offsetHeight, document.documentElement.offsetHeight),
+        };
+      });
+    } catch (e) {
+      return null;
+    }
+  }
+
+  if (!mainFrameSnapshot.url.startsWith('http'))
+    mainFrameSnapshot.url = 'http://playwright.snapshot/';
+  return {
+    viewportSize,
+    mainFrame: mainFrameSnapshot,
+  };
+}
+
+function takeSnapshot(guid: string, shadowAttribute: string, removeNoScript: boolean): any[] {
+  const win = window;
+  const doc = win.document;
+
+  const autoClosing = new Set(['AREA', 'BASE', 'BR', 'COL', 'COMMAND', 'EMBED', 'HR', 'IMG', 'INPUT', 'KEYGEN', 'LINK', 'MENUITEM', 'META', 'PARAM', 'SOURCE', 'TRACK', 'WBR']);
+  const escaped = { '&': '&amp;', '<': '&lt;', '>': '&gt;', '"': '&quot;', '\'': '&#39;' };
+
+  const escapeAttribute = (s: string): string => {
+    return s.replace(/[&<>"']/g, char => (escaped as any)[char]);
+  };
+  const escapeText = (s: string): string => {
+    return s.replace(/[&<]/g, char => (escaped as any)[char]);
+  };
+
+  const frames = new Map<string, Element>();
+  const chunks = new Map<string, string>();
+  const resources = new Set<string>();
+  let counter = 0;
+
+  const nextId = (): string => {
+    return guid + (++counter);
+  };
+
+  const resolve = (base: string, url: string): string => {
+    if (url === '')
+      return '';
+    try {
+      return new URL(url, base).href;
+    } catch (e) {
+      return url;
+    }
+  };
+
+  const addResource = (base: string, url: string): string => {
+    if (url.startsWith('javascript:'))
+      return '';
+    if (url.startsWith('data:') || !url)
+      return url;
+    resources.add(resolve(base, url));
+    return url;
+  };
+
+  const processSrcSet = (base: string, srcset: string): string => {
+    return srcset.split(',').map(src => {
+      src = src.trim();
+      const spaceIndex = src.lastIndexOf(' ');
+      if (spaceIndex === -1)
+        return addResource(base, src);
+      return addResource(base, src.substring(0, spaceIndex).trim()) + src.substring(spaceIndex);
+    }).join(',');
+  };
+
+  const processStyle = (base: string, style: string) => {
+    // TODO: this is wrong.
+    for (const match of style.matchAll(/(url\s*\(['"]?)([^)'"]*)(['"]?\))/g))
+      addResource(base, match[2]);
+  };
+
+  // TODO: handle CSSOM.
```

Comment:
What is the point in capturing style resources if we don't capture CSSOM?

Reply:
Yeah, I'll look into it.